_posts/2016-05-12-publish-subscribe-model-in-kafka.md

```diff
@@ -0,0 +1,123 @@
+
+## Queuing vs publish-subscribe
+
+Consumer groups is a key concept to explain why Kafka is more flexible and powerful than other messaging solutions like [RabbitMQ](https://www.rabbitmq.com/). Consumers are associated to consumer groups. If every consumer has the same consumer group, the topic's messages will be evenly load balanced between consumers; that's called queuing model. By contrast, if every consumer has a different consumer group, all the messages will be consumed in every client; that's called publish-subscribe model.
+
+You can have a mix of both approaches, having different logical consumer groups, for different needs, and several consumers inside of every group to increase throughput through parallelism.
+
+## Understanding our needs
+
+<img src="/assets/img/custom/blog/law_enforcement.png" alt="Law enforcement architecture" title="Law enforcement architecture" class="img img-center img-responsive style-screengrab">
+
+As we saw in previous posts the Items service publish messages into a Kafka topic called `item_deleted`. To define in which topic's partition will live that message Kafka provides [three alternatives](https://kafka.apache.org/090/javadoc/org/apache/kafka/clients/producer/Partitioner.html):
+
+> If a partition is specified in the record, use it
+
+>If no partition is specified but a key is present choose a partition based on a hash of the key
+
+>If no partition or key is present choose a partition in a round-robin fashion
+
+We'll use as a key the `item_id`. Consumers contained in different instances of the Law enforcement service are only interested in particular partitions, as they're keeping internal state for some items. Let's inspect different Kafka consumer implementations to see what is the more convenient for our use case.
+
+## Kafka Consumers
+
+There are three consumers in Kafka: [High level consumer](https://cwiki.apache.org/confluence/display/KAFKA/Consumer+Group+Example), [Simple Consumer](https://cwiki.apache.org/confluence/display/KAFKA/0.8.0+SimpleConsumer+Example) and [New Consumer](http://kafka.apache.org/090/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html)
+
+**Simple Consumer** is the lowest-level of them. It fits for our scenario as allows to "consume only a subset of the partitions in a topic in a process". However, as the documentation says:
+
+> The SimpleConsumer does require a significant amount of work not needed in the Consumer Groups:
+You must keep track of the offsets in your application to know where you left off consuming.
+You must figure out which Broker is the lead Broker for a topic and partition
+You must handle Broker leader changes
+
+If you read the code suggested for handling those concerns, you'll be quickly discouraged to use this consumer.
+
+**New Consumer** offers the right level of abstraction and allows us to subscribe to specific partitions. In the documentation they suggest as use case the following:
```

Comment:
In the documentation they suggest as use case the following ===> they suggest the following use case in the documentation